templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,185 @@
+    # from the registry. This is so a user can specify a different InstallDir than the
+    # directory where the current package is installed and the workaround will still work
+    # for the already installed package
+    $InstallDir = fetch_install_dir
+    if (Test-Path $temp_puppetres) {
+      Out-File -InputObject "Restoring puppetres.dll" -FilePath $Logfile -Append
+      Move-Item -Path $temp_puppetres -Destination "$InstallDir\puppet\bin\puppetres.dll"
+    }
+  }
+}
+
+# ************** Execution start **************
+$ErrorActionPreference = "Stop"
+$service_names=@(
+  "puppet",
+  "pxp-agent",
+  "mcollective"
+)
+$msi_arguments = "/qn /norestart /i `"$Source`" /l*vx+ `"$Logfile`" $InstallArgs"
+
+if (Test-Path "HKLM:\SOFTWARE\Puppet Labs") {
+  $msi_arguments += " ENVIRONMENT=`"$(puppet.bat config print --section agent environment)`""
+}
+if ($InstallDir) {
+  $msi_arguments += " INSTALLDIR=`"$InstallDir`""
+}
+if ($PuppetMaster) {
+  $msi_arguments += " PUPPET_MASTER_SERVER=`"$PuppetMaster`""
+}
+if ($PuppetStartType) {
+  $msi_arguments += " PUPPET_AGENT_STARTUP_MODE=`"$PuppetStartType`""
+}
+Out-File -InputObject $PID -FilePath $InstallScriptPIDFile
+try {
+  # Wait for the puppet run to finish
+  Out-File -InputObject "Waiting for puppet to stop, PID:$PuppetPID" -FilePath $Logfile -Append
+  try {
+    Wait-Process -Id $PuppetPID -Timeout 120
+  # We have to catch all exceptions here since the Error action preference of 'stop' will cause wait-process
+  # to actually throw [System.Management.Automation.ActionPreferenceStopException] rather than the actual
+  # exceptions. However, once inside the catch block the value of $_ is correctly set as the actual exception
+  } catch {
+    if ($_.Exception.GetType() -eq [Microsoft.PowerShell.Commands.ProcessCommandException]) {
+      Out-File -InputObject "Puppet was already finished, continuing..." -FilePath $Logfile -Append
+    } elseif ($_.Exception.GetType() -eq [System.TimeoutException]) {
+      Out-File -InputObject "ERROR: timed out waiting for puppet run to finish!" -FilePath $Logfile -Append
+      throw $_
+    } else {
+      Out-File -InputObject "ERROR: failed while trying to wait for puppet process with $($_.Exception.GetType())" -FilePath $Logfile -Append
+      throw $_
+    }
+  }
+  foreach($service in $service_names) {
+    $serv_exists = Get-Service -Name $service -ErrorAction SilentlyContinue
+    if ($serv_exists) {
+      Out-File -InputObject "Stopping $($service) before upgrade" -FilePath $Logfile -Append
+      Stop-Service $service
+    }
+  }
+  if ($UseLockedFilesWorkaround) {
+    $temp_puppetres = move_puppetres_dll
+  }
+  Out-File -InputObject "Beginning MSI installation with Arguments: $msi_arguments" -FilePath $Logfile -Append
+  Out-File -InputObject "****************************** Begin msiexec.exe output ******************************" -FilePath $Logfile -Append

```

Comment:
Something to watch out for here is that if the user specifies a service password, it may appear in this log in clear text

Reply:
Would the password appear because of the MSI output? or is there something in the script that would show that?

I was a little worried about that, but the current batch file was writing this log output. I can go either way with this, I was following the exact implementation of the batch file as much as I could.

Reply:
I believe the MSI logs it.  I mean we can punt the issue till later, but it shouldn't be forgotten